rfc: IPv6 Dual Stack Support for Cloud Foundry

peanball opened the pull request:
This PR adds the RFC "IPv6 Dual Stack Support for Cloud Foundry".

For easier viewing, you can see [the full RFC](https://github.com/sap-contributions/cf-community/blob/rfc-ipv6-in-cf/toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md) as preview.

rkoster commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,200 @@
+3. Support of dual stack networking, i.e. IPv4 and IPv6 at the same time. This is primarily for a smoother migration because the IPv6 changes can then be additive.
+4. IPv6 is additive to IPv4 functionality
+
+The ultimate future goal would be to move to pure IPv6 communication but is not a current priority and is not covered in this proposal.
```
IPv4?

peanball replied:
IPv6 networks allow interoperability, e.g. via NAT64 and similar. IPv6-only is still intended as "option" on how to use Cloud Foundry. We will clarify this.

peanball replied:
Modified that statement to a more realistic goal.

ameowlia commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,200 @@
+
+### Windows Support
+
+Windows support is currently not in the scope of this RFC.
```
I think windows should be part of this RFC. If we are going to add IPv6 support we should add it for all apps not just linux.

peanball replied:
Based on the comment by @erobertson, it seems like there is no particular specific requirement / deviation for Windows.

We will still require support with the implementation and testing of IPv6 / dual stack related changes on Windows.

peanball replied:
@erobertson, I've updated the statement. Could you please cross check that what is written now is correct?

peanball replied:
@ebrobertson, I've updated the statement. Could you please cross check that what is written now is correct?

ebroberson replied:
Looks good to me!

ameowlia commented:
Hi @peanball, Thanks for putting this together. 

I know there is a lot here, and a lot in App Runtime Platform's domain. My main thoughts are:

* This is going to touch a lot of things. I sure we will learn about more things as we go.
* I agree with your comment that this should be backwards compatible. All of the defaults for all new properties should default to the "old way".
* I think windows should be in scope, unless there is a good reason not to include it. We maintain two OSes we should support both.

peanball commented:
@ameowlia, thanks for the feedback. 

Regarding windows: we will need support here in describing what needs to be touched as we (SAP) are not using Windows as target.

jochenehret commented:
Hi @peanball ,
to adopt IPv6 in cf-deployment and other projects, we will need an update of **bosh-bootloader**. The tool must be able to configure AAAA records for the load balancer ingress addresses. Not sure what changes are needed for the outgoing traffic. The cloud-configs must be updated with the new CIDRs. There are certainly more changes necessary, is anyone already working on that?

For **cf-deployment**, we will have to update our test bbl environments when bosh-bootloader is ready. The "staging" and "running" security groups must be updated, but that should be moderate effort. There will be certainly more changes necessary...

Best Regards,

  Jochen.

mariash commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,200 @@
+
+The following sections outline the changes necessary for various core components of  a typical CF foundation.
+
+### BOSH
```
Doesn't BOSH support ipv6 already if CPI supports it? - https://bosh.io/docs/guide-ipv6-on-vsphere/

peanball replied:
IPv6 yes, dual stack not yet.

mariash commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,200 @@
+
+### Diego
+
+Currently, the Silk CNI is used to create an overlay network, but also to assign IP addresses for the application containers. This assignment is done from a centrally managed pool of IP addresses.
```
Was Silk CNI work specifically not included in this RFC?

peanball replied:
Please also note the extended section on Silk that was added by @plamen-bardarov

mariash commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,200 @@
+
+Network traffic from and to application containers needs to be controlled, also for IPv6:
+
+1. For inter-container communication, Application Security Groups (ASGs) need to support IPv6.
```
I think majority of work would be to support dual-stack. For CNI wrapper plugin which configures ASGs IP address is an arbitrary string.

peanball replied:
I've changed the wording to "need to add support for IPv6", which implies dual stack. Everything in this RFC is about adding IPv6 to the existing IPv4 and thus reaching dual-stack support.

Do you think we need to be more explicit here?

mariash commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,200 @@
+Network Policies that are stored in [the policy-server-api which is part of cf-networking-release](https://github.com/cloudfoundry/cf-networking-release/blob/develop/docs/08-policy-server-api.md).
+
+Network policies may reference IP address ranges as destination. Destinations are stored as strings. The underlying IP ranges are also [strings with a start and end address](https://github.com/cloudfoundry/cf-networking-release/blob/aa25a3f96dc21e8a3e25db10f741e4f678c78464/src/code.cloudfoundry.org/policy-server/api/api.go#L37-L49). The data is ultimately serialized JSON that is stored in a datab ...
+
```
Same here with the dual-stack. If we want to roll this out for existing deployments dual-stack should be supported for backwards compatibility. 

Most of the components only set up IP address right now.

peanball replied:
I will add some sentence in the beginning that highlights that this is about adding IPv6 and thus reaching dual stack.

mariash commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,200 @@
+routing-api and tcp-router need to be verified, and if necessary extended, to support IPv6 addresses in addition to IPv4 addresses.
+
+## Other Topics
+
```
I think another thing probably to mention is that all components that have API (Cloud Controller, Policy Server, UAA, Credhub, BBS, Rep, Mysql) should be able to accept IPv6 connections. It is probably already managed by the http modules they use, unless some explicitly listen on 127.0.0.1.

mariash replied:
Also, BOSH DNS should resolve IPv6 addresses depending on the client protocol.

mariash replied:
Another thing, CATS should have end to end tests for IPv6 (ASG rules, App listening on IPv6, etc)

peanball replied:
I've added the point about "listeners" generally being IPv6 / dual stack capable, also including BOSH DNS. 

CATs and bbl (thanks @jochenehret) are now mentioned as well.

peanball commented:
Jochen wrote:
>to adopt IPv6 in cf-deployment and other projects, we will need an update of bosh-bootloader. The tool must be able to configure AAAA records for the load balancer ingress addresses. Not sure what changes are needed for the outgoing traffic. The cloud-configs must be updated with the new CIDRs. There are certainly more changes necessary, is anyone already working on that?

@fmoehler, could you support me in extend on that?

ameowlia commented:
> @ameowlia, thanks for the feedback.
> 
> Regarding windows: we will need support here in describing what needs to be touched as we (the current authors of this RFC) are not using Windows as target.

I've reached out to @ebroberson, our resident Windows expert, to comment more about what changes will be needed for windows.

stephanme commented:
For cf-deployment I think it makes sense to define an (experimental) ops-file for enabling IPv6 early so that the different components have an integration point for their settings.

Probably it makes sense to add a validation pipeline for IPv6 as soon as bbl and bosh supports it. Similar to the noble validation pipeline.

ebroberson commented:
Windows enables IPv6 by default as of Windows Server 2008, so I don't think we'll need anything specific there.

beyhan commented:
> Windows enables IPv6 by default as of Windows Server 2008, so I don't think we'll need anything specific there.

@ebroberson what about the following parts:
https://github.com/sap-contributions/cf-community/blob/639791ad22e3d82ac3f861ad4a9741bde0295ce9/toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md?plain=1#L101-L103 and https://github.com/sap-contributions/cf-community/blob/rfc-ipv6-in-cf/toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md?plain=1#L55?

ebroberson commented:
@beyhan I was referring to anything _else_ that might be Windows-specific on the stemcells / vms.  Windows itself already supports dual-stack with IPv6, so I don't believe we'll need anything further other than what the RFC already specifies.

peanball commented:
@stephanme, @rkoster, @ameowlia, @ebroberson, @jochenehret, @mariash, @beyhan, I have addressed the feedback provided so far. Thank you for your feedback!

We might need to write a bit more about the bosh-bootstrap / bbl environments. Help here would be appreciated, as I've not worked with bbl environments or bosh-bootstrap yet.

jpalermo commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,269 @@
+  - name: diego_cells_ipv6
+    subnets:
+      - az: z1
+        cloud_properties:
+          # assign non-overlapping `/80` cidrs from `range`
+          ipv6_prefix_delegation_size: 80
```
cloud_properties are generally reserved for things that are NOT bosh features. A feature intended for only a single CPI to be able to enable something specific in the particular IaaS.

I'm don't have an accurate view of all the pieces we're dealing with here, but if it's a general purpose ipv6 thing, it should be part of the core network properties, not down in cloud_properties.

It may be possible to add new upper level properties without breaking existing CPI APIs, but if not, it's not a huge deal to decide to bump the CPI API version from 2 to 3. We've only done it a single time, so it sounds like a big deal, but really all the CPIs declare their version and the Bosh director asks them which version they support for every call. So implementing new CPI version 3 things isn't a big deal if needed.

peanball replied:
Hi @jpalermo thanks for your feedback!

This is a general feature for IPv6. Prefix delegation is a basic feature of the  general IPv6 networking stacks. While not all hyperscalers currently support prefix delegation on their VMs, most already do.

This is a fundamental feature for enabling IPv6 address delegation, i.e. "IPv6-address per Diego LRP instance container".

CPI version 3 sounds reasonable for adding dual stack support and the various IPv6 bits.

I'm not an expert on BOSH's core network properties.

Would it be possible that you either point me to the right place, or adapt the above example to something that would fit into the core network properties?

beyhan replied:
This has been addressed with https://github.com/cloudfoundry/community/pull/1077/commits/0eec79e146c4c5058a4abd791aaaea3e3cac0535. @fmoehler could you please confirm.

maxmoehl commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,269 @@
+
+BOSH is used to provision and configure the VMs that run the components of Cloud Foundry. The following features are needed to support the goals of this proposal:
+
+1. Assignment of IPv6 addresses in addition to IPv4, ideally to the same NIC
+2. Assignment of IPv6 CIDR ranges (prefix delegation), e.g. assign a `/80` address, while avoiding overlapping. This is relevant for the use case of address delegation to application workloads hosted on Diego cells.
```
If we can assign prefixes, assigning a `/128` prefix should be equivalent to assigning an IPv6 address (but I'm not in the details, there might be some subtle differences) so maybe we "only" need prefix assignment?

Regarding prefixes for diego: there might be a benefit to using different prefixes for the cell itself and app containers. That way you can define ACLs for all apps / cells via a single prefix. This would require BOSH to be capable of assigning multiple IPv6 prefixes to a single VM.

a-hassanin replied:
I would separate IP Address assignment and Prefix assignment here 'to not force' a consumer to specify a prefix. At the end, if an IPv6 address is specified without a prefix, the IaaS Provider; like AWS; can always apply the default.

Regarding assigning multiple IPv6 prefixes on a single VM, after the discussion with @peanball, it seems it is not required to be in scope for this RFC, but indeed, it is something we can add later on if required.

peanball replied:
The "not required" statement is not set in stone. We are trying to figure out with this RFC, how to best do it.

I agree that assigning prefixes (which under the hood you always do anyway!) would be good. IF needed, we could have a "lazy" mode where you define just the address and not a CIDR, and it automatically assigns as `/128`. But those are finer implementation details.

The general suggestion, which I find interesting and worth including, is that we could have distinct address ranges for the Diego cells themselves, and for the container workloads on each Diego cell.

This actually works into @rkoster's question about the prefix delegation size, which should be more dynamic as it is at the moment. 

It's really only Diego at the moment that requires prefix delegation, as it does something with those delegated addresses.
Other deployments require a single address and are good. She the "delegation prefix" is then a single address for all deployments that are not Diego. And that for Diego we would have one with a single address, and one with a larger delegated prefix.

maxmoehl replied:
Multi-address allows for quite some interesting setups. For example, when combined with bpm you could have separate prefixes for all metric/logging/... components which other components cannot talk to (e.g. on a gorouter VM, rsyslog would get a "logging" IP, but gorouter itself would not). This would reduce the impact of a compromised service as it can only talk to its peers and not just everyone. It's questionable whether you would actually do something like this in our context as it is quite some effort to get right.

plamen-bardarov replied:
> The general suggestion, which I find interesting and worth including, is that we could have distinct address ranges for the Diego cells themselves, and for the container workloads on each Diego cell.

From the diego/silk perspective, having a dedicated IPv6 prefix per cell exclusively for application containers is beneficial as it prevents conflicts when assigning IPv6 addresses, since the host-local IPAM plugin only checks its local state files and does not verify existing addresses on network interfaces. It also enhances network separation, making it easier to define ACLs if needed for apps/cells

beyhan replied:
My personal suggestion would be to focus on the minimal scope which implements the desired support for IPv6. Multi-address is also possible with IPv4 and IIRC we tried similar things as described in @maxmoehl [comment](https://github.com/cloudfoundry/community/pull/1077#discussion_r1982732684) but had many issues with that and decided against it. It could be that with IPv6 it is easier to get this setup work but It looks to me as topic for the possible future improvement section of this RFC. 

If I got the discussions right another open point is to update the RFC with mentioning support for distinct address ranges for the Diego cells themselves, and for the container workloads on each Diego cell. 

@peanball we didn't start the FCP because of this open discussion. If this resolves until Tuesday next week most probably we can start the FCP.

PlamenDoychev replied:
I do agree with the above statement from @beyhan, if not documented we should document that we do support distinct address ranges for the Diego cells themselves, and for the container workloads.

May-be it is also worth mentioning that multi-address is out of scope for this iteration.

peanball replied:
I'm adding a sentence that a single CIDR will will be assigned. This is also in line with the BOSH config sample that defines a single prefix delegation size and subnet. Anything more elaborate will need a different configuration mechanism.

None of this precludes us to extend in the future however.

rkoster commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
```diff
@@ -0,0 +1,269 @@
+            - sg-abcd...
+          subnet: subnet-12345
+        # assign non-overlapping `/80` cidrs from `range`
+        ipv6_prefix_delegation_size: 80
```
should this be specified at the cloud-config level, or at the deployment level? Is there a use-case for mixing prefixes? As in diego-cells get a /80 but cloud-controller won't need such a prefix.

peanball replied:
Yes. Absolutely.

peanball replied:
I've now responded to a comment where the topic is related: https://github.com/cloudfoundry/community/pull/1077#discussion_r1982018440

fmoehler replied:
i put in a small snippet adding this as another option: https://github.com/cloudfoundry/community/pull/1077/commits/4b8f22cf9d8b0bd37ffef41d5332df7272684642

peanball replied:
I'm adding emphasis on this 2nd option, pointing it out as preference. Also I've added a warning that we need to be careful when the prefix size changes.

beyhan commented:
We discussed this RFC during the TOC meeting on 11th of March and decided to start the FCP with the goal to accept the RFC.

peanball commented on toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md:
````diff
@@ -0,0 +1,312 @@
+
+```
+
+The preference is to use the deployment-based expression for the prefix size, as it provides better control to the deployment and does not require changes to the cloud config.
````
It's not the network range but the "step size" for divvying up the range so the smaller segments can be assigned to the resulting VMs in the deployment.

tl;dr: The deployment defines how many instances there will be that need to fit into the overall CIDR provided to the respective BOSH network. This informs the "step size" that can be used to ensure non-overlapping addresses.

The deployment should be agnostic to the environment, but only up to a point. Depending on your environment, you might have a larger or smaller IPv6 CIDR available to give to the overall Diego cell network and to subdivide.

Say you want to lay out your network in the `2001:db8:4e04:D139:0000:0000:0000:0000/96` CIDR.

If you want to maximize the address space for 15 Diego cells, you then might want to assign to each Diego cell a `/100` CIDR from the `/96` Diego network CIDR:
* `2001:db8:4e04:D139:0000:0000:0000:0000/100` (according to our examples: reserved for the deployment, DNS, etc.)
* `2001:db8:4e04:D139:0000:0000:1000:0000/100` - cell 1
* `2001:db8:4e04:D139:0000:0000:2000:0000/100` - cell 2
* `2001:db8:4e04:D139:0000:0000:3000:0000/100` - cell 3
* `2001:db8:4e04:D139:0000:0000:4000:0000/100` - cell 4
* `2001:db8:4e04:D139:0000:0000:5000:0000/100` - cell 5
* `2001:db8:4e04:D139:0000:0000:6000:0000/100` - cell 6
* `2001:db8:4e04:D139:0000:0000:7000:0000/100` - cell 7
* `2001:db8:4e04:D139:0000:0000:8000:0000/100` - cell 8
* `2001:db8:4e04:D139:0000:0000:9000:0000/100` - cell 9
* `2001:db8:4e04:D139:0000:0000:a000:0000/100` - cell 10
* `2001:db8:4e04:D139:0000:0000:b000:0000/100` - cell 11
* `2001:db8:4e04:D139:0000:0000:c000:0000/100` - cell 12
* `2001:db8:4e04:D139:0000:0000:d000:0000/100` - cell 13
* `2001:db8:4e04:D139:0000:0000:e000:0000/100` - cell 14
* `2001:db8:4e04:D139:0000:0000:f000:0000/100` - cell 15

Each cell will get ≈268 million IPv6 addresses from their `/100` CIDR range.

But if you decide that you want to have 350 Diego cells, you would pick a smaller "step" of e.g. `/108` that would give each cell ≈1 million and could fit 4096 cells.


But if your initial CIDR for the Diego network is a `/64`, you have other considerations. Because I don't want to prescribe a size for Diego, or for the minimum CIDR to give to Diego, it might be best for the deployment to decide how bit a slice to expect from the overall CIDR for each of the VMs.

rkoster replied:
Re-reading the full paragraph, and it all makes sense. Thanks for the clarification.
